pinot-perf/src/main/java/org/apache/pinot/perf/LazyDataList.java

```diff
@@ -0,0 +1,163 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.perf;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import org.apache.pinot.spi.data.readers.GenericRow;
+
+
+public class LazyDataList implements List<GenericRow> {
```

Comment:
This is not a list. Two calls to get(index) with the same index may return different values. Instead of creating this break of abstraction, I suggest to create an IterableRecordReader (or maybe rewrite GenericRowRecordReader to accept an iterable instead of a List).

Reply:
I refactored this piece because it turned out that, while it doesn't need to be a list, segment creation does multiple passes over data, so generator must be able to rewind and recreate data properly.

Reply:
Why ugly? It should be the same, right? You just need to initialize random with a known value and reset to it each time you need to rewind.